Experiment with ArrayByteBufferPool performance

gregw opened the pull request:
No overall size accounting
reserved buffer release always checks max memory
released buffers check max memory 1% of the time.
only a single thread can check memory at once.
single pass through buckets so no looping forever.

gregw committed:
Experiment with ArrayByteBufferPool

No overall size accounting
reserved buffer release always checks max memory
released buffers check max memory 1% of the time.
only a single thread can check memory at once.
single pass through buckets so no looping forever.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -267,46 +255,39 @@ private void release(RetainedBucket bucket, Pool.Entry<RetainableByteBuffer> ent
         // We have enough space for this entry, pool it.
         if (entry.release())
+        {
+            if (buffer instanceof Buffer b && b.use() % 100 == 1)
```
If `b.use()` returns a negative value, the result of `%100` will also be negative. You should test ` == 0` to avoid that.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -267,46 +255,39 @@ private void release(RetainedBucket bucket, Pool.Entry<RetainableByteBuffer> ent
-        // Account the excess at most for the bucket capacity.
-        return Math.min(capacity, newMemory - maxMemory);
+        long max = direct ? _maxDirectMemory : _maxHeapMemory;
+        if (max <= 0 || !_evictor.compareAndSet(false, true))
```
I don't get what ensuring that only a single thread can evict is buying is.

gregw replied:
We don't want multiple threads all calculating the excess and evicting it at the same time, as they could evict  too much memory.
This is especially true during initial phase when most buffers are new and we always check the maxMemory from reserve.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -267,46 +255,39 @@ private void release(RetainedBucket bucket, Pool.Entry<RetainableByteBuffer> ent
+            long memory = getMemory(direct);
+            long excess = memory - max;
+            if (excess > 0)
+                evict(excess, direct);
```
I assume you removed the logic about `evict()` returning false to remove the current entry because the odds are too great that this would happen despite the eviction?

gregw replied:
Because we have already added the released buffer when we check for max memory, so we can't avoid churn.... but at least we would only churn on 1% and then only 1 thread.

lorban commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ArrayByteBufferPoolTest.java:
```diff
@@ -73,7 +72,7 @@ public void testMaxMemoryEviction()
         assertThat(pool.getDirectByteBufferCount(), lessThan((long)buffers.size()));
         assertThat(pool.getDirectMemory(), greaterThan(0L));
-        assertThat(pool.getDirectMemory(), lessThanOrEqualTo(40L));
+        assertThat(pool.getDirectMemory(), lessThan(120L));
```
This seems to indicate that we can now badly overshoot. I'm wondering if the configurable memory limit makes any more sense given this.

gregw replied:
We can overshoot a bit for a while, but I don't think it is badly, as the next thread to evict will again attempt to get it down all the way.
This was just a lazy test change and we really need to do a probabilistic test to match the algorithm

gregw replied:
@lorban I've improved the test a bit.  We still overshoot, but you can see it is not by much

gregw committed:
Experiment with ArrayByteBufferPool

updates from review

gregw commented:
@lorban can you perf test this both in JMH and overall so we can see if this makes a difference, or do we need to look for a better approach.

lorban commented:
I think this is the best we can do, given what we expect from ABBP.

Could you please just add the following test to the benchmark:
```java
@Benchmark
    @BenchmarkMode({Mode.Throughput})
    public void fastPathAcquireRelease()
    {
        RetainableByteBuffer buffer = pool.acquire(65535, true);
        buffer.release();
    }
```

and modify the threads count to 32 in the `main` method?

lorban commented:
FYI here's a perf run of this branch:
https://jenkins.webtide.net/job/load_testing/job/jetty-profiler-12.0.x/128/artifact/core/target/reports/CoreHandlerPerfTest_testNoGzipAsync/server/load-master-2/perf.hlog.html

Perf of the fast path is on par with 12.0.6.

gregw commented:
@lorban what was the "hiccup" in the middle of that perf run?
Does it always do that?

gregw committed:
JMH updates

gregw commented:
@sbordet thoughts?

lorban commented:
@gregw that hiccup is caused by async-profiler; disabling it makes the hiccup go away, and yes: it always dips around the 60th second.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -415,14 +390,10 @@ public long getHeapMemory()
 
     private long getMemory(boolean direct)
```
Please check getAvailableMemory() that is now a duplicate of getMemory() and similar methods for get[Heap|Direct](Available)Memory().

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -415,14 +390,10 @@ public long getHeapMemory()
-        memory.addAndGet(-amount);
+        long size = 0;
+        for (RetainedBucket bucket : direct ? _direct : _indirect)
+            size += (long)bucket.getPool().getIdleCount() * bucket._capacity;
```
bucket.getCapacity()

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -267,46 +255,39 @@ private void release(RetainedBucket bucket, Pool.Entry<RetainableByteBuffer> ent
         // We have enough space for this entry, pool it.
         if (entry.release())
+        {
+            if (buffer instanceof Buffer b && b.use() % 100 == 0)
```
Calling use() after release() is racy (another thread can pick the buffer, use it and released it, and now 2 threads are calling use().

gregw committed:
updates from review

sbordet committed:
Fixed comments.
Fixed call to recordEvict().
Removed unused methods.
Method getAvailable*Memory() no longer JMX-enabled, as they are the same as get*Memory().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
